HDDS-3149. Make o3fs support set and get acl

captainzmc opened the pull request:
## What changes were proposed in this pull request?

BasicOzoneFileSystem currently lacks an interface to set and get acls through HDFS. Supplement with this PR.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3149

## How was this patch tested?

Ut has been added

lokeshj1703 commented on hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java:
```diff
@@ -322,6 +330,37 @@ public void makeQualified(FileStatus status, URI uri, Path path,
   }
 
+  @Override
+  public Map<String, List<OzoneAcl>> getAcl(String keyName) throws IOException {
```
We can rename the api to getAcls and setAcls since we are dealing with multiple acls.

lokeshj1703 commented on hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneFileSystem.java:
```diff
@@ -703,6 +710,78 @@ public String toString() {
+    adapter.setAcl(pathToKey(path), aclSpecToString(aclSpec));
+  }
+
+  public String toAclString(AclEntry aclEntry) {
```
I think we are converting AclEntry to form which can be parsed by OzoneAcl. Would it make sense to move this function to OzoneAcl class?

captainzmc commented:
> @captainzmc Thanks for working on this! The changes look good to me. Please find my comments below.
> 
> 1. There are few test failures. Can you please look at that?
> 2. Can you also add a unit test for the added apis in TestOzoneFileSystem?
> 3. There are other acl based apis in FileSystem class. Can you also create jiras for those?

Hi @lokeshj1703 Very thanks for your review and you suggestion is good.
1. By reopen this PR, run the UT, the second run was successful. Look at the previous error, should be caused by other places. Maybe there need another JIRA to fix ut.
2. I will consider increase ut in TestOzoneFileSystem. 
3. We've tested the BaseOzoneFile API before and found a lot of unsupported content([HDDS-3048](https://issues.apache.org/jira/browse/HDDS-3048)). Currently, we have plans to improve this part of the content(such as removeAcl and so on).

captainzmc commented:
Hi @lokeshj1703 I had fixed review issue and added ut. Could you please review this pr again?
    I made some changes according to your suggestion. About increase UT TestOzoneFileSystem, I found that the current o3fs doesn't support to do that. Current access to o3fs is based on volume and bucket, they need to be created in advance. And when we enable acl, the volume creation permission is denied, because o3fs API cannot currently manipulate volumes and buckets, and  permissions also cannot be set by o3fs API.  I added some UT to TestReadWriteStatistics to test this PR.

lokeshj1703 commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/OzoneAcl.java:
```diff
@@ -340,4 +342,37 @@ public OzoneAcl setAclScope(AclScope scope) {
     DEFAULT;
   }
+
+  public String toAclString(AclEntry aclEntry) {
```
Can we add documentation for these functions that String returned can be parsed using OzoneAcl#parseAcl?

lokeshj1703 commented on hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java:
```diff
@@ -322,6 +330,38 @@ public void makeQualified(FileStatus status, URI uri, Path path,
+    ozoneClient.getObjectStore().getAcl(obj).forEach(ozoneAcl -> {
+      if (ozoneAcl.getType().equals(ACLIdentityType.USER)) {
+        userResult.add(ozoneAcl);
+      } else if (ozoneAcl.getType().equals(ACLIdentityType.GROUP)) {
```
Sorry for identifying this late, do we also need to check the other ACLIdentityType here like ACLIdentityType.WORLD etc.?

lokeshj1703 commented on hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneFileSystem.java:
```diff
@@ -703,6 +708,43 @@ public String toString() {
+   *         (default outcome).
+   */
+  @Override
+  public void setAcl(Path path, List<AclEntry> aclSpec) throws IOException {
```
The AclEntry can be of type MASK, OTHER. I think we should map these into OzoneAcl as well, but dont know the correct mapping here. cc @xiaoyuyao

xiaoyuyao replied:
OzoneAcl currently does not support MASK acl. But I don't think we should put the Ozone user/group acl into owner and group string here. Instead, they should be mapped to AclStatus as List<AclEntry>.

captainzmc commented:
Hi @lokeshj1703, I see this part of the implementation in ofs (hdds-2991) as well, the implementations of the two acls should be consistent, and there should be common methods on both sides. To avoid any overlap, I will temporarily close the pr until hdds-2991 is merged.
